### PR TITLE
feat: Add validation rules to create account

### PR DESCRIPTION
**Pull request by MaximeDion-Work**
GitHub Issue: #

## PR Type

What kind of change does this PR introduce?

- Feature

## Description

Adding more complex example of fluent validation

## PR Checklist 
Please check if your PR fulfills the following requirements:

- [ ] ~~Interface members are XML documented~~
- [ ] ~~Documentation (XML or comments) has been added and/or existing documentation has been updated~~
- [ ] ~~[Architecture documents](./doc/Architecture.md) have been updated~~
- [x] Tested on all relevant platforms

### CODE REVIEW

File: src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems [hunk 1]
```diff
@@ -9,6 +9,9 @@
     <Import_RootNamespace>ApplicationTemplate.Views</Import_RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)Behavior\FormattingTextBoxBehavior.Android.cs" />
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems:12]: I don't think we should be adding this behavior to the template. You don't actually need it for your validation features.

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs [hunk 2]
```diff
@@ -20,19 +43,111 @@ public string Email
...
 			RuleFor(x => x.Email).NotEmpty().EmailAddress();
-			RuleFor(x => x.Password).NotEmpty();
+			RuleFor(x => x.PhoneNumber).NotEmpty().Length(14).WithMessage(localizer["CreateAccount_PhoneNumberValidation"]); // "(000) 000-0000".Length = 14
+			RuleFor(x => x.SecondaryPhoneNumber).Must(phonenumber =>
+			{
+				if (phonenumber.IsNullOrEmpty())
+				{
+					return true;
+				}
+				else if (phonenumber.Length == 14)
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:111]: there are multiple ways of writing a phone number. You shouldn't check the string length, but rather its content.

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs [hunk 3]
```diff
@@ -20,19 +43,111 @@ public string Email
...
+				else
+				{
+					return false;
+				}
+
+			}).WithMessage(localizer["CreateAccount_PhoneNumberValidation"]);
+			RuleFor(x => x.PostalCode).NotEmpty().Length(7).WithMessage(localizer["CreateAccount_PostalValidation"]); // "A1A 1A1".Length = 7
+			RuleFor(x => x.DateOfBirth).NotEmpty().Must(dateOfBirth =>
+			{
+				var diff = DateTime.Now.Year - dateOfBirth.DateTime.Year;
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:124]: you should be comparing DateTimeOffsets, not DateTimes

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs [hunk 4]
```diff
@@ -20,19 +43,111 @@ public string Email
...
+					return true;
+				}
+				else
+				{
+					return false;
+				}
+
+			}).WithMessage(localizer["CreateAccount_PhoneNumberValidation"]);
+			RuleFor(x => x.PostalCode).NotEmpty().Length(7).WithMessage(localizer["CreateAccount_PostalValidation"]); // "A1A 1A1".Length = 7
+			RuleFor(x => x.DateOfBirth).NotEmpty().Must(dateOfBirth =>
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:122]: Can you create at least 1 extension method for a custom rule?
Like `.MustBe18OrOlder()`

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs [hunk 5]
```diff
@@ -20,19 +43,111 @@ public string Email
...
+			{
+				var diff = DateTime.Now.Year - dateOfBirth.DateTime.Year;
+				return diff > 18;
+			}).WithMessage(localizer["CreateAccount_DateOfBirthValidation"]);
+			RuleFor(x => x.Password).NotEmpty().Must(password =>
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:127]: When you have 2 or more extension chained, can you put each extension method on its own line, starting with the "."

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs (hunk 2 above)
Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:111]: but in our case the the phone number is strongly formatted, only numbers and other characters are added automatically.

File: src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems (hunk 1 above)
Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems:12]: But it make the sample more realistic for Phone Number and Postal Code fields

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs (hunk 4 above)
Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs]: In my extension method, how do I get access to the value that is being test?

image

File: src/app/ApplicationTemplate.UWP/Views/Content/Authentication/CreateAccountPage.xaml [hunk 6]
```diff
@@ -44,19 +143,67 @@
...
+						<!-- Label -->
+						<TextBlock Text="Favorite Donuts (pick at least 3)"
+								   Foreground="Black"
+								   FontSize="16" />
+						<TextBlock Text="{Binding Form, Converter={StaticResource debug}, Mode=TwoWay}"
+								   Foreground="Black"
+								   FontSize="16" />
+
+						<!-- List -->
+						<ListView ItemsSource="{Binding Donuts}"
```
Comment (MaximeDion-Work) [inline src/app/ApplicationTemplate.UWP/Views/Content/Authentication/CreateAccountPage.xaml:180]: FavoriteDonuts is always null for some reason. Debug converter on Form shows that's null and when breaking into the validation rule show that its null

File: src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems (hunk 1 above)
Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems]: Added comment in xaml to suggest implementing some form of text formating, this is a bit out of scope for the feature.

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs (hunk 2 above)
Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs]: specified expected formatting in the titles of phone numbers

**Comment by MaximeDion-Work** [comment]
@jeanplevesque Added the extension method for 18 or older.

File: src/app/ApplicationTemplate.Shared/Presentation/Extensions/AbstractValidatorExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Collections.Generic;
+using FluentValidation;
+
+namespace Presentation
+{
+	public static class AbstractValidatorExtensions
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Presentation/Extensions/AbstractValidatorExtensions.cs:7]: The class name doesn't quite match the extended type.

File: src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems [hunk 8]
```diff
@@ -23,4 +23,7 @@
     <Content Include="$(MSBuildThisFileDirectory)Assets\Image_Example.scale-300.png" />
     <Content Include="$(MSBuildThisFileDirectory)Assets\Image_Example.scale-400.png" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="$(MSBuildThisFileDirectory)Behavior\" />
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared.Views/ApplicationTemplate.Shared.Views.projitems:27]: Remove this if empty

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs [hunk 9]
```diff
@@ -20,19 +44,138 @@ public string Email
...
+			var myPassword = string.Empty;
+
+			RuleFor(x => x.FirstName).NotEmpty();
+			RuleFor(x => x.LastName).NotEmpty();
+
+			RuleFor(x => x.Email)
+				.NotEmpty()
+				.EmailAddress();
+
+			RuleFor(x => x.PhoneNumber)
```
Comment (MatFillion) [inline src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:109]: We should count the digits, not all characters, no?

Comment (jeanplevesque) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:109]: that's true

Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:109]: Ideally we would use a fixed formatting behavior like (000) 000-0000. I think this gives a good enough representation of fluent validation.

Comment (jeanplevesque) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:109]: For this sample, we could probably should remove all formatting. The important part is the data.

Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:109]: Unless you really want it too be looking for digit only.

Comment (MaximeDion-Work) [reply src/app/ApplicationTemplate.Shared/Presentation/Authentication/CreateAccountFormViewModel.cs:110]: Done

**Review by jeanplevesque — APPROVED** [review]

**Pull request merged**